prov/rxm/src/rxm_conn.c

```diff
@@ -1361,19 +1403,161 @@ static void *rxm_conn_progress(void *arg)
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to get MSG_EP EQ WAIT_FD %d\n", ret);
+		goto exit;
+	}
+	fds.events = POLLIN;
+
+	while (1) {
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
+		if (rxm_ep->msg_cq) {
+			ofi_ep_lock_release(&rxm_ep->util_ep);
+			return FI_SUCCESS;
 		}
-		ret = rxm_conn_handle_event(rxm_ep, entry);
-		if (ret)
+		again = fi_trywait(rxm_fabric->msg_fabric, &fid, 1);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
+
+		if (!again) {
+			fds.revents = 0;
+
+			ret = poll(&fds, 1, -1);
+			if (OFI_UNLIKELY(ret == -1)) {
+				if (errno == EINTR)
+					continue;
+				FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+					"Select error %d, closing CM thread\n",
+					errno);
+				goto exit;
+			}
+		}
+		if (rxm_conn_auto_progress_eq(rxm_ep, entry))
 			goto exit;
-		memset(entry, 0, RXM_MSG_EQ_ENTRY_SZ);
 	}
+exit:
+	return -1;
+}
+
+/* Atomic auto progress of EQ and CQ */
+static int rxm_conn_atomic_progress_eq_cq(struct rxm_ep *rxm_ep,
+					  struct rxm_msg_eq_entry *entry)
+{
+	struct rxm_fabric *rxm_fabric;
+	struct fid *fids[2];
+	struct pollfd fds[2];
+	int again;
+	int ret;
+
+	rxm_fabric = container_of(rxm_ep->util_ep.domain->fabric,
+				  struct rxm_fabric, util_fabric);
+
+	fids[0] = &rxm_ep->msg_eq->fid;
+	ret = fi_control(fids[0], FI_GETWAIT, (void *) &fds[0].fd);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to get MSG_EP EQ WAIT_FD %d\n", ret);
+		goto exit;
+	}
+	fds[0].events = POLLIN;
+
+	fids[1] = &rxm_ep->msg_cq->fid;
+	fds[1].fd = rxm_ep->msg_cq_fd;
+	fds[1].events = POLLIN;
+	memset(entry, 0, RXM_MSG_EQ_ENTRY_SZ);
+
+	while(1) {
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
+		again = fi_trywait(rxm_fabric->msg_fabric, fids, 2);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
+
+		if (!again) {
+			fds[0].revents = 0;
+			fds[1].revents = 0;
+
+			ret = poll(fds, 2, -1);
+			if (OFI_UNLIKELY(ret == -1)) {
+				if (errno == EINTR)
+					continue;
+				FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+					"Select error %d, closing CM thread\n",
+					errno);
+				goto exit;
+			}
+		}
+		if (again || fds[0].revents & POLLIN) {
+			if (rxm_conn_auto_progress_eq(rxm_ep, entry))
+				goto exit;
+		}
+		if (again || fds[1].revents & POLLIN)
+			rxm_ep_progress(&rxm_ep->util_ep);
+	}
+exit:
+	return -1;
+}
+
+static void *rxm_conn_atomic_progress(void *arg)
+{
+	struct rxm_ep *rxm_ep = container_of(arg, struct rxm_ep, util_ep);
+	struct rxm_msg_eq_entry *entry;
+
+	assert(rxm_ep->msg_eq);
+	entry = calloc(1, RXM_MSG_EQ_ENTRY_SZ);
+	if (!entry) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to allocate memory!\n");
+		return NULL;
+	}
+
+	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
+	       "Starting CM conn thread with atomic AUTO_PROGRESS\n");
+
+	/* To reduce overhead, a unique progress function that progresses
+	 * only the EQ is used until the CQ has been bound to the EP */
+	if (rxm_conn_atomic_progress_eq(rxm_ep, entry) < 0)
+		goto exit;
+
+	rxm_conn_atomic_progress_eq_cq(rxm_ep, entry);
```

Comment:
Could negate if() statement and move up to avoid goto and exit label

Reply:
Sounds good, I updated the PR. Thanks.